File: 776-n-ary-tree-postorder-traversal/solution.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val,List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    public List<Integer> postorder(Node root) {
        List<Integer> result = new LinkedList<>();
        Stack<Node> stack = new Stack<>(); stack.push(root);
        while (!stack.empty() && root != null) {
            Node node = stack.pop();
            result.add(0, node.val);
            for (Node n : node.children) stack.push(n);
        }
        return result;
    }

    public List<Integer> postorder2(Node root) {
        List<Integer> result = new LinkedList<>();
        if (root != null) {
            List<Integer> res = new LinkedList<>();
            for (Node n : root.children) res.addAll(postorder(n));
            result.addAll(res);
            result.add(root.val);
        }
        return result;
    }
}

File: 776-n-ary-tree-postorder-traversal/solution.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
public:
    vector<int> postorder(Node* root) {
        vector<int> result, vec, temp;
        if (root) {
            for (auto child : root->children) {
                temp = postorder(child);
                vec.insert(vec.end(), temp.begin(), temp.end());
            }
            result.insert(result.end(), vec.begin(), vec.end());
            result.push_back(root->val);
        }
        return result;
    }
    
    vector<int> postorder2(Node* root) {
        stack<Node*> s; s.push(root);
        vector<int> result;
        while (!s.empty() && root) {
            Node * n = s.top(); s.pop();
            result.insert(result.begin(), n->val);
            for (auto child : n->children) s.push(child);
        }
        return result;
    }
};
